hw/ip_templates/alert_handler/alert_handler_reg.core.tpl

```diff
@@ -0,0 +1,27 @@
+CAPI=2:
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+name: ${instance_vlnv("lowrisc:ip:alert_handler_reg:0.1")}
+description: "Auto-generated alert handler register sources"
+virtual:
+  # TODO: Check if the generated files actually maintain a stable public API.
```

Comment:
hmmm i see. so these are called `ip_interfaces` if they remain fixed despite the other top specific parts.
Does this mean if `reg_pkg` for example changes based on top, it should also be lumped into `opentitan`?

Reply:
The idea is: IP blocks like `lc_ctrl` need some functionality from any instance of `alert_handler`, but they don't care if it is `top_earlgrey_alert_handler`, the `top_englishbreakfast_alert_handler`, or any other instance. In case of `lc_ctrl`, `lc_ctrl.sv` needs (among other things) the constant `alert_handler_reg_pkg::N_ESC_SEV`, which will always be present in all implementations of alert_handler. (Yes, the constant will have a different value depending on the toplevel, but it'll always be there.)

If you have such a dependency, you can now depend on a virtual core (aka an interface), just like on a real core.

To *declare* the interface, the `virtual:` keyword is used, which effectively gives the core a second name (or even a third one). To *use* the interface, simply write the virtual core name as dependency, as if it were a normal core.

https://github.com/olofk/fusesoc/pull/524 has some more information in the commit message.

Once we're happy with things here I'll send around an email update or a quick silicon meeting update to ensure everybody knows about these new constructs.

To answer your other question:
>  Does this mean if reg_pkg for example changes based on top, it should also be lumped into opentitan?

A templated core needs to be in the `lowrisc:opentitan:top_earlgrey_*` namespace if it doesn't expose a stable interface. An interface (for our purpose) is the module header (input/output ports), parameters/types/enums in a package, etc. In all cases, only the identifiers matter for the definition of an interface: the values assigned to them don't matter.

With regard to the reg_pkg packages: They don't expose a fully stable interface, but a subset of the interface is stable. We seem to be only using this stable subset right now (e.g. `alert_handler_reg_pkg::N_ESC_SEV` in `lc_ctrl.sv`), but it's a bit fragile and needs to be looked into at some point.

Reply:
Yeah I agree that it would be good to have a couple of slides to explain this mechanism to the wider Si audience to explain how this is supposed to be used correctly.